dashboards-observability/public/components/visualizations/charts/maps/treemaps.tsx

```diff
@@ -3,57 +3,85 @@
-      branchvalues: 'total',
-      labels,
-      parents,
-      domain: { x: [0.52, 1] },
-      values: [65, 14, 12, 10, 2, 6, 6, 1, 4],
-      textinfo: 'label+value+percent parent+percent entry',
-      outsidetextfont: { size: 20, color: '#377eb8' },
-      marker: { line: { width: 2 } },
-      pathbar: { visible: false },
-    },
-  ];
-  const finalLayout = {
-    annotations: [
-      {
-        showarrow: false,
-        text: 'branchvalues: <b>remainder</b>',
-        x: 0.25,
-        xanchor: 'center',
-        y: 1.1,
-        yanchor: 'bottom',
-      },
+export const TreeMap = ({ visualizations, layout, config }: any) => {
+  const {
+    data,
+    metadata: { fields },
+  } = visualizations.data.rawVizData;
+  const { dataConfig = {}, layoutConfig = {} } = visualizations?.data?.userConfigs;
+
+  const labelField =
+    dataConfig?.valueOptions &&
+    dataConfig?.valueOptions.labelField &&
+    !isEmpty(dataConfig?.valueOptions.labelField)
+      ? dataConfig?.valueOptions.labelField[0]
+      : fields[fields.length - 1];
+
+  const parentField =
+    dataConfig?.valueOptions &&
+    dataConfig?.valueOptions.parentField &&
+    !isEmpty(dataConfig?.valueOptions.parentField)
+      ? dataConfig?.valueOptions.parentField[0]
+      : null;
+
+  const valueField =
+    dataConfig?.valueOptions &&
+    dataConfig?.valueOptions.valueField &&
+    !isEmpty(dataConfig?.valueOptions.valueField)
+      ? dataConfig?.valueOptions.valueField[0]
+      : fields[0];
+
+  if (
+    isEmpty(data[labelField.name]) ||
+    isEmpty(data[valueField.name]) ||
+    (!isNull(parentField) && isEmpty(data[parentField.name])) ||
+    isEqual(labelField, parentField) ||
+    indexOf(NUMERICAL_FIELDS, valueField.type) < 0
+  )
+    return <EmptyPlaceholder icon={visualizations?.vis?.iconType} />;
+
+  const treemapData = useMemo(() => {
+    let labelsArray, parentsArray, valuesArray;
+
+    if (parentField === null) {
+      labelsArray = [...data[labelField.name]];
+      parentsArray = [...Array(labelsArray.length).fill('')];
+      valuesArray = [...data[valueField.name]];
+    } else {
+      const uniqueParents = uniq(data[parentField.name]);
+      labelsArray = [...data[labelField.name], ...uniqueParents];
+      parentsArray = [...data[parentField.name], ...Array(uniqueParents.length).fill('')];
+      valuesArray = [...data[valueField.name], ...Array(uniqueParents.length).fill(0)];
+    }
+
+    return [
       {
-        showarrow: false,
-        text: 'branchvalues: <b>total</b>',
-        x: 0.75,
-        xanchor: 'center',
-        y: 1.1,
-        yanchor: 'bottom',
+        type: 'treemap',
+        labels: labelsArray,
+        parents: parentsArray,
+        values: valuesArray,
+        textinfo: 'label+value+percent parent+percent entry',
       },
-    ],
+    ];
+  }, [data, labelField, valueField, parentField]);
+
+  const mergedLayout = {
+    ...layout,
+    ...(layoutConfig.layout && layoutConfig.layout),
```

Comment:
The spread operator can create a problem. The code in brackets is going to produce a boolean (false) if `layoutConfig.layout` is empty or null. Again, there should be a key in the object that can take this boolean as a value. For `||`, we are sure that the statement always returns the last value if everything fails.